src/tactic/ring_exp.lean

```diff
@@ -0,0 +1,1430 @@
+
+lemma simple_pf_sum_zero {p p' : α} : p = p' → p + 0 = p' := by simp
+
+lemma simple_pf_prod_one {p p' : α} : p = p' → p * 1 = p' := by simp
+
+lemma simple_pf_prod_neg_one {α} [ring α] {p p' : α} : p = p' → p * -1 = - p' := by simp
+
+lemma simple_pf_var_one (p : α) : p ^ 1 = p := by simp
+
+lemma simple_pf_exp_one {p p' : α} : p = p' → p ^ 1 = p' := by simp
+
+/--
+  Give a simpler, more human-readable representation of the normalized expression.
+
+  Normalized expressions might have the form `a^1 * 1 + 0`,
+  since the dummy operations reduce special cases in pattern-matching.
+  Humans prefer to read `a` instead.
+  This tactic gets rid of the dummy additions, multiplications and exponentiations.
+-/
+meta def ex.simple : Π {et : ex_type}, ex et → ring_exp_m (expr × expr)
+| sum pps@(ex.sum pps_i p (ex.zero _)) := do
+  (p_p, p_pf) ← p.simple,
+  prod.mk p_p <$> mk_app_csr ``simple_pf_sum_zero [p.pretty, p_p, p_pf]
+| sum (ex.sum pps_i p ps) := do
+  (p_p, p_pf) ← p.simple,
+  (ps_p, ps_pf) ← ps.simple,
+  prod.mk
+    <$> mk_add [p_p, ps_p]
+    <*> mk_app_csr ``sum_congr [p.pretty, p_p, ps.pretty, ps_p, p_pf, ps_pf]
+| prod (ex.prod pps_i p (ex.coeff _ ⟨⟨1, 1, _, _⟩⟩)) := do
+  (p_p, p_pf) ← p.simple,
+  prod.mk p_p <$> mk_app_csr ``simple_pf_prod_one [p.pretty, p_p, p_pf]
+| prod pps@(ex.prod pps_i p (ex.coeff _ ⟨⟨-1, 1, _, _⟩⟩)) := do
+  ctx ← get_context,
+  match ctx.info_b.ring_instance with
+  | none := prod.mk pps.pretty <$> pps.proof_term
+  | (some ringi) := do
+    (p_p, p_pf) ← p.simple,
+    prod.mk
+      <$> lift (mk_app ``has_neg.neg [p_p])
+      <*> mk_app_class ``simple_pf_prod_neg_one ringi [p.pretty, p_p, p_pf]
+  end
+| prod (ex.prod pps_i p ps) := do
+  (p_p, p_pf) ← p.simple,
+  (ps_p, ps_pf) ← ps.simple,
+  prod.mk
+    <$> mk_mul [p_p, ps_p]
+    <*> mk_app_csr ``prod_congr [p.pretty, p_p, ps.pretty, ps_p, p_pf, ps_pf]
+| base (ex.sum_b pps_i ps) := ps.simple
+| exp (ex.exp pps_i p (ex.coeff _ ⟨⟨1, 1, _, _⟩⟩)) := do
+  (p_p, p_pf) ← p.simple,
+  prod.mk p_p <$> mk_app_csr ``simple_pf_exp_one [p.pretty, p_p, p_pf]
+| exp (ex.exp pps_i p ps) := do
+  (p_p, p_pf) ← p.simple,
+  (ps_p, ps_pf) ← in_exponent $ ps.simple,
+  prod.mk
+    <$> mk_pow [p_p, ps_p]
+    <*> mk_app_csr ``exp_congr [p.pretty, p_p, ps.pretty, ps_p, p_pf, ps_pf]
+| et ps := prod.mk ps.pretty <$> ps.proof_term
+
+/--
+  Performs a lookup of the atom `a` in the list of known atoms,
+  or allocates a new one.
+
+  If `a` is not definitionally equal to any of the list's entries,
+  a new atom is appended to the list and returned.
+  The index of this atom is kept track of in the second inductive argument.
+
+  This function is mostly useful in `resolve_atom`,
+  which updates the state with the new list of atoms.
+-/
+meta def resolve_atoms (a : expr) : list atom → ℕ → ring_exp_m (atom × list atom)
+| [] n := let atm : atom := ⟨a, n⟩ in pure (atm, [atm])
+| bas@(b :: as) n := do
+  (lift $ is_def_eq a b.value >> pure (b , bas)) <|> do
```

Comment:
Back to my comment toward the top. `is_def_eq` takes an optional reducibility parameter. The default is `semireducible`. You should take an optional flag in the `ring_exp` front end (`ring_exp!`) and thread the flag to here. If the flag is present, use `semireducible`. Otherwise use `reducible`.

Reply:
Done!